rllib/core/models/torch/primitives.py

```diff
@@ -56,3 +62,106 @@ def __init__(
+                except for the output).
+            output_activation: The activation function to use for the output layer.
+            output_dim: The output dimensions are [BATCH, output_dim]. We append a final
+                convolutional layer depth-only filters that is flattened and a final
+                linear layer to achieve this.
+        """
+        super().__init__()
+
+        assert len(input_dims) == 3
+        assert isinstance(output_dim, int)
+        assert filter_specifiers is not None, "Must provide filter specifiers."
+
+        filter_layer_activation = get_activation_fn(
+            filter_layer_activation, framework="torch"
+        )
+
+        output_activation = get_activation_fn(output_activation, framework="torch")
+
+        layers = []
+        # Create some CNN layers
+        core_layers = []
+
+        # Add user-specified hidden convolutional layers first
+        width, height, in_depth = input_dims
+        in_size = [width, height]
+        for out_depth, kernel, stride in filter_specifiers:
+            padding, out_size = same_padding(in_size, kernel, stride)
+            # TODO(Artur): Inline SlimConv2d
+            core_layers.append(
+                SlimConv2d(
+                    in_depth,
+                    out_depth,
+                    kernel,
+                    stride,
+                    padding,
+                    activation_fn=filter_layer_activation,
+                )
+            )
+            in_depth = out_depth
+            in_size = out_size
+
+        core_cnn = nn.Sequential(*core_layers)
+        layers.append(core_cnn)
+
+        # Append a last convolutional layer of depth-only filters to be flattened.
+
+        # Get info of the last layer of user-specified layers
+        [width, height] = in_size
+        _, kernel, stride = filter_specifiers[-1]
+
+        in_size = (
+            int(np.ceil((width - kernel[0]) / stride)),
+            int(np.ceil((height - kernel[1]) / stride)),
+        )
+        padding, _ = same_padding(in_size, (1, 1), (1, 1))
+        # TODO(Artur): Inline SlimConv2d or use TorchCNN here
+        layers.append(
+            SlimConv2d(
+                in_depth,
+                1,
+                (1, 1),
+                1,
+                padding,
+                activation_fn=filter_layer_activation,
+            )
+        )
+        layers.append(nn.Flatten())
+
+        # Add a final linear layer to make sure that the outputs have the correct
+        # dimensionality.
+        layers.append(nn.Linear(int(width) * int(height), output_dim))
+        if output_activation is not None:
+            layers.append(output_activation())
+
+        # Create the cnn that potentially includes a flattened layer
+        self.cnn = nn.Sequential(*layers)
+
+    def forward(self, x):
+        # Permute b/c data comes in as [B, dim, dim, channels]:
+        inputs = x.permute(0, 3, 1, 2)
```

Comment:
This is why spec checking is required :) How would I know if the input is in the correct format. NICE !